<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>convert-variations</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rel="stylesheet" href="main.css" type="text/css" />
</head>

<body class="info">

<center>
<h2><A HREF="RSAT_home.cgi">RSAT</A> - <a href='convert-variations_form.cgi'>convert-variations</a> manual</h2>

</center>

<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>

<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#VERSION">VERSION</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#AUTHORS">AUTHORS</a></li>
  <li><a href="#CATEGORY">CATEGORY</a></li>
  <li><a href="#INPUT-DATA">INPUT DATA</a>
    <ul>
      <li><a href="#Genome-Variant-Format-GVF">Genome Variant Format (GVF)</a></li>
      <li><a href="#Variant-Call-Format-VCF">Variant Call Format (VCF)</a></li>
      <li><a href="#RSAT-variation-Format-varBed">RSAT variation format (varBed)</a></li>
    </ul>
  </li>
  <li><a href="#OUTPUT-FORMAT">OUTPUT FORMAT</a></li>
  <li><a href="#OPTIONS">OPTIONS</a></li>
      <ul>
      <li><a href="#Variants-to-be-converted">Variants to be converted </a></li>
      <li><a href="#Input-format">Input format</a></li>
      <li><a href="#Output-format">Output format</a></li>

    </ul>
  <li><a href="#CONTACT">CONTACT</a></li>
</ul>

<h1 id="NAME">NAME</h1>
  <p>convert-variations</p>
<h1 id="VERSION">VERSION</h1>
  <p>2.0</p>
<h1 id="DESCRIPTION">DESCRIPTION</h1>
  <p>Performs inter-conversions between different formats of polymorphic variations.</p>
<h1 id="AUTHORS">AUTHORS</h1>
<dl>
  <dt id="Walter Santana-Garcia">Walter Santana-Garcia</dt>
<dd>

</dd>
  <dt id="Jacques van Helden">Jacques van Helden</dt>
<dd>

</dd>
  <dt id="Alejandra-Medina-Rivera">Alejandra Medina-Rivera</dt>
<dd>

</dd>
</dl>

<h1 id="CATEGORY">CATEGORY</h1>
<dl>
<dt id="util">Genetic variations</dt>
</dl>

<!-- <h1 id="USAGE">USAGE</h1>

<pre><code> variation-scan [-i sequence_file] -m matrix_file -bg backgournd_file [-calc_distrib] [-o outputfile] [-v #] [...]</code></pre>

<h2 id="Example">Example</h2> -->

<h1 id="INPUT-DATA">INPUT DATA</h1>
<p>Genome Variant Format (GVF), Variant Call Format (VCF) and RSAT variation format (varBed).</p>

<h2 id="Genome-Variant-Format-GVF">Genome Variant Format (GVF)</h2>

<p>&quot;The Genome Variant Format (GVF) is a type of GFF3 file with additional pragmas and attributes specified. The GVF format has the same nine column tab delimited format as GFF3 and all of the requirements and restrictions specified for GFF3 apply to the GVF specification as well.&quot; (quoted from the Sequence Ontology)</p>

<p>http://www.sequenceontology.org/resources/gvf_1.00.html</p>

<p>A GVF file starts with a header providing general information about the file content: format version, date, data source, length of the chromosomes / contigs covered by the variations.</p>

<pre><code> ##gff-version 3
 ##gvf-version 1.07
 ##file-date 2014-09-21
 ##genome-build ensembl GRCh38
 ##species http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=9606
 ##feature-ontology http://song.cvs.sourceforge.net/viewvc/song/ontology/so.obo?revision=1.283
 ##data-source Source=ensembl;version=77;url=http://e77.ensembl.org/Homo_sapiens
 ##file-version 77
 ##sequence-region Y 1 57227415
 ##sequence-region 17 1 83257441
 ##sequence-region 6 1 170805979
 ##sequence-region 1 1 248956422
 ## [...]</code></pre>

<p>This header is followed by the actual description of the variations, in a column-delimited format compying with the GFF format.</p>

<pre><code> Y       dbSNP   SNV     10015   10015   .       +       .       ID=1;variation_id=23299259;Variant_seq=C,G;Dbxref=dbSNP_138:rs113469508;allele_string=A,C,G;evidence_values=Multiple_observations;Reference_seq=A
 Y       dbSNP   SNV     10146   10146   .       +       .       ID=2;variation_id=26647928;Reference_seq=C;Variant_seq=G;evidence_values=Multiple_observations,1000Genomes;allele_string=C,G;Dbxref=dbSNP_138:rs138058540;global_minor_allele_frequency=0|0.0151515|33
 Y       dbSNP   SNV     10153   10153   .       +       .       ID=3;variation_id=21171339;Reference_seq=C;Variant_seq=G;evidence_values=Multiple_observations,1000Genomes;allele_string=C,G;Dbxref=dbSNP_138:rs111264342;global_minor_allele_frequency=1|0.00229568|5
 Y       dbSNP   SNV     10181   10181   .       +       .       ID=4;variation_id=47159994;Reference_seq=C;Variant_seq=G;evidence_values=1000Genomes;allele_string=C,G;Dbxref=dbSNP_138:rs189980076;global_minor_allele_frequency=0|0.00137741|3</code></pre>

<p>The last column contains a lot of relevant information, but is not very easy to read. We should keep in mind that this format was initially defined to describe generic genomic features, so all the specific attributes come in the last column (description).</p>

<!-- <p>For this reasons, we developed a custom tab-delimited format to store variations, which we call <i>varBed</i> (see description below).</p> -->

<h2 id="Variant-Call-Format-VCF">Variant Call Format (VCF)</h2>

<p>http://en.wikipedia.org/wiki/Variant_Call_Format</p>

<p>This format was defined for the 1000 genomes project. It is no longer maintained. The converter supports it merely for the sake of backwards compatibility.</p>

<h2 id="RSAT-variation-Format-varBed">RSAT variation format (varBed)</h2>

<p>Tab-delimited format with a specific column order, used as input by <a href='retrieve-variation-seq_form.cgi'><i>retrieve-variation-seq</i></a>.</p>

<p>This format presents several advantages for scanning variations with matrices.</p>
<h1 id="OUTPUT-FORMAT">OUTPUT FORMAT</h1>
<p>A tab delimited file on selected output format.</p>


<h1 id="OPTIONS">OPTIONS</h1>

<dl>


<dt id="Variants-to-be-converted "><b>Variants to be converted </b></dt>
<dd>

<p>Variation data that will be converted, supported formats: GVF, VCF or varBed.</p>

</dd>

<dt id="Input-format"><b>Input format </b></dt>
<dd>

<p> Variation format of the input data, supported formats: GVF, VCF or varBed.</p>

</dd>

<dt id="Output-format"><b>Output format</b></dt>
<dd>

<p>Variation format of the desired output data, supported formats: GVF, VCF or varBed</p>

</dd>

</dl>

  <h1 id="CONTACT">CONTACT</h1>
  <dl>
    <dd>For further inquiries, please contact Jacques van Helden (<i>Jacques.van-Helden@univ-amu.fr</i>) or
      <a href= 'https://twitter.com/RSATools'>Ask a question to the RSAT team</a>!</dd>
  </dl>


</body>

</html>
