<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>variation-info</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rel="stylesheet" href="main.css" type="text/css" />
</head>

<body class="info">

<center>
<h2><A HREF="RSAT_home.cgi">RSAT</A> - <a href='variation-info_form.cgi'>variation-info</a> manual</h2>

</center>

<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>

<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#VERSION">VERSION</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#AUTHORS">AUTHORS</a></li>
  <li><a href="#CATEGORY">CATEGORY</a></li>
  <li><a href="#INPUT-DATA">INPUT DATA</a>
    <ul>
      <li><a href="#Genomic-coordinate-file">Genomic coordinate file</a></li>
      <li><a href="#Variation-ID-list">Variation ID list</a></li>
    </ul>
  </li>
  <li><a href="#OUTPUT-FORMAT">OUTPUT FORMAT</a></li>
  <li><a href="#OPTIONS">OPTIONS</a></li>
      <ul>
      <li><a href="#Organism">Organism </a></li>
      <li><a href="#Variations-IDs-or-genomic-regions-of-interest">Variations IDs or genomic regions of interest </a></li>
      <li><a href="#Input-format">Input format</a></li>

    </ul>
  <li><a href="#CONTACT">CONTACT</a></li>
</ul>

<h1 id="NAME">NAME</h1>
  <p>variation-info</p>
<h1 id="VERSION">VERSION</h1>
  <p>2.0</p>
<h1 id="DESCRIPTION">DESCRIPTION</h1>
  <p>Taking as input variation IDs (rs numbers) or regions in a given genome, <i>variation-info</i> retrieves variant information associated to the queries in varBed format.</p>
<h1 id="AUTHORS">AUTHORS</h1>
<dl>
  <dt id="Walter Santana-Garcia">Walter Santana-Garcia</dt>
<dd>

</dd>
  <dt id="Jacques van Helden">Jacques van Helden</dt>
<dd>

</dd>
  <dt id="Alejandra-Medina-Rivera">Alejandra Medina-Rivera</dt>
<dd>

</dd>
</dl>

<h1 id="CATEGORY">CATEGORY</h1>
<dl>
<dt id="util">Genetic variations</dt>
</dl>

<!-- <h1 id="USAGE">USAGE</h1>

<pre><code> variation-scan [-i sequence_file] -m matrix_file -bg backgournd_file [-calc_distrib] [-o outputfile] [-v #] [...]</code></pre>

<h2 id="Example">Example</h2> -->

<h1 id="INPUT-DATA">INPUT DATA</h1>

<h2 id="Genomic-coordinate-file">Genomic coordinate file</h2>

<p>A genomic coordinate file in bed format can be used as input to retrieve overlapping variants from ensembl databases. The tool only takes into account the 3 first columns of the bed file, which specify the genomic coordinates.</p>

<p><b>Note</b> (from Jacques van Helden): the UCSC genome browser adopts a somewhat inconsistent convention for start and end coordinates: the start position is zero-based (first nucleotide of a chromosome/scaffold has coordinate 0), but the end position is considered not included in the selection. This is equivalent to have a zero-based coordinate for the start, and a 1-base coordinate for the end.</p>

<b>Example of bed file</b>

<pre><code> chr1   3473041 3473370
 chr1   4380371 4380650
 chr1   4845581 4845781
 chr1   4845801 4846260</code></pre>

<p>The definition of the BED format is provided on the UCSC Genome Browser web site (http://genome.ucsc.edu/FAQ/FAQformat#format1).</p>

<p>This tool only takes into account the 3 first columns, which specify the genomic coordinates.</p>

<dl>

<dt id="chrom">1. chrom</dt>
<dd>

<p>The name of the chromosome (e.g. chr3, chrY, chr2_random) or scaffold (e.g. scaffold10671).</p>

</dd>
<dt id="chromStart">2. chromStart</dt>
<dd>

<p>The starting position of the feature in the chromosome or scaffold. For RSAT programs, the first base in a chromosome is numbered 1 (this differs from the UCSC-specific zero-based notation for the start).</p>

<p><b>Note</b> from Jacques van Helden: the UCSC genome browser adopts a somewhat inconsistent convention for start and end coordinates: the start position is zero-based (first nucleotide of a chromosome/scaffold has coordinate 0), and the end position is considered not included in the selection. This is equivalent to have a zero-based coordinate for the start, and a 1-base coordinate for the end. We find this representation completely counter-intuitive, and we herefore decided to adopt a &quot;normal&quot; convention, where:</p>

<dl>

<dt id="start-and-end-position-represent-the-first-and-last-positions-included-in-the-region-of-interest">start and end position represent the first and last positions <i>included</i> in the region of interest.</dt>
<dd>

</dd>
<dt id="start-and-end-positions-are-provided-in-one-based-notation-first-base-of-a-chromosome-or-contig-has-coordinate-1">start and end positions are provided in one-based notation (first base of a chromosome or contig has coordinate 1).</dt>
<dd>

</dd>
</dl>

</dd>
<dt id="chromEnd">3. chromEnd</dt>
<dd>

<p>The ending position of the feature in the chromosome or scaffold.</p>

</dd>
</dl>

<h2 id="Variation-ID-list">Variation ID list</h2>

<p>An ID (dbSNP) list of variation in one column.</p>

<h1 id="OUTPUT-FORMAT">OUTPUT FORMAT</h1>

<p>varBed format is a tab delimited file that facilitates access to relevant variant information. The file includes the following columns:</p>

<dl>

<dt id="chr">1) chr</dt>
<dd>

<p>Chromosome number (without &quot;chr&quot;).</p>

</dd>
<dt id="start">2) start</dt>
<dd>

<p> Start position of the variation.</p>

</dd>
<dt id="end">3) end</dt>
<dd>

<p>End position of the variation.</p>

</dd>
<dt id="strand">4) strand</dt>
<dd>

<p>Strand were the variation is found.</p>

</dd>
<dt id="id">5) id</dt>
<dd>

<p>Variant ID, rs number.</p>

</dd>
<dt id="ref">6) ref</dt>
<dd>

<p>Reference allele.</p>

</dd>
<dt id="alt">7) alt</dt>
<dd>

<p>Alternative allele(s).</p>

</dd>
<dt id="so_term-validate">8) so_term </dt>
<dd>

<p> Sequence Ontology term of the variant. </p>

</dd>

 <dt id="alt">9) validate</dt>
  <dd>

  <p>Validation status of the variant: 1 if it had evidence, 0 otherwhise.</p>

  </dd>

<dt id="minor_allele_freq">10) minor_allele_freq</dt>
<dd>

<p>Frequency of the alternative allele.</p>

</dd>

</dl>

<!--<h1 id="SEE-ALSO">SEE ALSO</h1>

<h2 id="download-ensembl-genome">download-ensembl-genome</h2>

<p><i>retrieve-variation-seq</i> uses the sequences downloaded from Ensembl using the tool <i>download-ensembl-genome</i>.</p>

<h2 id="download-ensembl-variations">download-ensembl-variations</h2>

<p><i>retrieve-variation-seq</i> uses variation coordinates downloaded from Ensembl using the tool <i>download-ensembl-variations</i>.</p>

<h2 id="variation-scan">variation-scan</h2>

<p>Scan variation sequences with one or several position-specific scoring matrices.</p>

<h1 id="WISH-LIST">WISH LIST</h1> -->

<h1 id="OPTIONS">OPTIONS</h1>

<dl>


<dt id="Organism"><b>Organism </b></dt>
<dd>

<p>Name of the organism where the queries will be searched.</p>

</dd>

<dt id="Variations-IDs-or-genomic-regions-of-interest"><b>Variations IDs or genomic regions of interest  </b></dt>
<dd>

<p> Input data, provided either as a list of variants IDs (dbSNPs) or as set of genomic coordinates (BED file).</p>

</dd>

<dt id="Input-format"><b>Input format</b></dt>
<dd>

<p>Format of the supplied queries, supported: bed or id.</p>

</dd>

</dl>

  <h1 id="CONTACT">CONTACT</h1>
  <dl>
    <dd>For further inquiries, please contact Jacques van Helden (<i>Jacques.van-Helden@univ-amu.fr</i>) or
      <a href= 'https://twitter.com/RSATools'>Ask a question to the RSAT team</a>!</dd>
  </dl>


</body>

</html>
